File: backend/models/gate_event.py
"""
GateKeeper – Modello GateEvent
================================
Registra ogni evento di passaggio dalla porta.

Un evento viene creato dal Raspberry ogni volta che:
  - un tag RFID UHF viene letto dal reader (oggetto passa)
  - un MAC BLE viene rilevato/perso vicino alla porta (utente esce/entra)

Questi eventi sono il cuore del sistema: l'app li mostra in tempo reale
e la logica smart (notifiche, alert) si basa su di essi.
"""

import uuid
from datetime import datetime

from sqlalchemy import String, DateTime, ForeignKey, Text, func
from sqlalchemy.orm import Mapped, mapped_column, relationship

from database import Base


class GateEvent(Base):
    """
    Tabella: gate_events

    Colonne:
        id: UUID primario.
        home_id: FK alla casa.
        user_id: FK utente associato all'evento (può essere NULL se non identificato).
        rfid_object_id: FK oggetto RFID coinvolto (NULL se evento solo-utente).
        event_type: Tipo evento (rfid_detected | ble_detected | manual).
        direction: Direzione (IN = entra in casa | OUT = esce di casa).
        timestamp: Quando è avvenuto l'evento.
        note: Nota opzionale (es. "ombrello dimenticato").
        raw_data: JSON grezzo del payload hardware (per debug).
    """

    __tablename__ = "gate_events"

    id: Mapped[str] = mapped_column(
        String(36), primary_key=True, default=lambda: str(uuid.uuid4())
    )
    home_id: Mapped[str] = mapped_column(
        String(36), ForeignKey("homes.id"), nullable=False
    )

    # L'utente può essere NULL se il sistema non riesce ad associare l'evento
    user_id: Mapped[str | None] = mapped_column(
        String(36), ForeignKey("users.id"), nullable=True
    )

    # L'oggetto RFID è NULL per eventi BLE-only (rilevamento utente)
    rfid_object_id: Mapped[str | None] = mapped_column(
        String(36), ForeignKey("rfid_objects.id"), nullable=True
    )

    # rfid_detected = tag RFID letto dal reader
    # ble_detected  = telefono rilevato/perso via BLE
    # manual        = evento creato manualmente dall'utente nell'app
    event_type: Mapped[str] = mapped_column(String(30), nullable=False)

    # IN = oggetto/utente entra in casa
    # OUT = oggetto/utente esce di casa
    direction: Mapped[str] = mapped_column(String(3), nullable=False)  # "IN" | "OUT"

    timestamp: Mapped[datetime] = mapped_column(
        DateTime, server_default=func.now(), index=True  # index per query veloci per data
    )

    note: Mapped[str | None] = mapped_column(Text, nullable=True)

    # Payload grezzo dal reader (JSON stringa) – utile per debug hardware
    # TODO: strutturalo meglio con una tabella separata se il volume cresce
    raw_data: Mapped[str | None] = mapped_column(Text, nullable=True)

    # ── Relazioni ──────────────────────────────────────────────────────────────
    home: Mapped["Home"] = relationship("Home", back_populates="gate_events")
    user: Mapped["User | None"] = relationship("User", back_populates="gate_events")
    rfid_object: Mapped["RfidObject | None"] = relationship("RfidObject", back_populates="gate_events")

    def __repr__(self) -> str:
        return (
            f"<GateEvent id={self.id!r} type={self.event_type!r} "
            f"direction={self.direction!r} ts={self.timestamp!r}>"
        )

File: app/lib/core/models/gate_event.dart
// ============================================================
// GateEvent — modello dati per un evento porta rilevato dal sistema
// ============================================================
//
// Ogni volta che il reader RFID + BLE rilevano un movimento
// (entrata o uscita) il Raspberry genera un GateEvent.
// Questo modello rappresenta quell'evento nell'app Flutter.
//
// TODO: allineare i campi con backend/app/models/schemas.py → EventSchema

/// Direzione del passaggio rilevato dalla porta.
enum EventDirection {
  /// Oggetto/utente è uscito di casa.
  exit,

  /// Oggetto/utente è rientrato in casa.
  entry,
}

/// Livello di priorità/importanza dell'evento per le notifiche.
enum EventSeverity {
  /// Evento normale, solo informativo.
  info,

  /// Qualcosa da tenere d'occhio (es. bambino uscito senza adulto).
  warning,

  /// Evento critico (es. oggetto sensibile portato fuori senza utente noto).
  critical,
}

/// Rappresenta un singolo evento rilevato dalla porta GateKeeper.
///
/// Contiene informazioni su chi è passato (utente identificato via BLE),
/// cosa è passato (oggetti rilevati via RFID) e quando.
///
/// Gli eventi sono la "memoria" del sistema: vengono mostrati
/// nella schermata Events in ordine cronologico.
///
/// Esempio:
/// ```dart
/// final event = GateEvent.fromJson(json);
/// final dir = event.direction == EventDirection.exit ? 'uscito' : 'rientrato';
/// print('${event.userDisplayName} è $dir');
/// ```
class GateEvent {
  /// ID univoco dell'evento (UUID generato dal backend).
  final String id;

  /// Momento in cui è stato rilevato l'evento (ora del Raspberry).
  final DateTime timestamp;

  /// Direzione del passaggio: uscita o entrata.
  final EventDirection direction;

  /// ID dell'utente associato all'evento tramite BLE.
  /// Null se nessun telefono BLE è stato rilevato vicino alla porta
  /// (potrebbe essere un estraneo o l'utente senza telefono).
  final String? userId;

  /// Nome visualizzato dell'utente (denormalizzato dal backend per comodità).
  /// Null se userId è null.
  final String? userDisplayName;

  /// Lista dei tag RFID degli oggetti passati con questo evento.
  /// Può essere vuota se solo l'utente è passato senza oggetti rilevati.
  final List<String> rfidTags;

  /// Lista dei nomi degli oggetti (denormalizzati dal backend).
  final List<String> objectNames;

  /// Livello di severità: info, warning, critical.
  final EventSeverity severity;

  /// Messaggio di notifica generato dal backend
  /// (es. "Simone è uscito senza ombrello ☔").
  final String? notificationMessage;

  const GateEvent({
    required this.id,
    required this.timestamp,
    required this.direction,
    this.userId,
    this.userDisplayName,
    required this.rfidTags,
    required this.objectNames,
    required this.severity,
    this.notificationMessage,
  });

  /// Costruisce un [GateEvent] da JSON ricevuto dall'API.
  factory GateEvent.fromJson(Map<String, dynamic> json) {
    return GateEvent(
      id: json['id'] as String,
      timestamp: DateTime.parse(json['timestamp'] as String),
      direction: json['direction'] == 'exit'
          ? EventDirection.exit
          : EventDirection.entry,
      userId: json['user_id'] as String?,
      userDisplayName: json['user_display_name'] as String?,
      rfidTags: List<String>.from(json['rfid_tags'] as List? ?? []),
      objectNames: List<String>.from(json['object_names'] as List? ?? []),
      severity: _severityFromString(json['severity'] as String? ?? 'info'),
      notificationMessage: json['notification_message'] as String?,
    );
  }

  /// Serializza l'evento in Map.
  /// Nota: di solito gli eventi li crea il backend, non il client.
  Map<String, dynamic> toJson() {
    return {
      'id': id,
      'timestamp': timestamp.toIso8601String(),
      'direction': direction.name,
      if (userId != null) 'user_id': userId,
      if (userDisplayName != null) 'user_display_name': userDisplayName,
      'rfid_tags': rfidTags,
      'object_names': objectNames,
      'severity': severity.name,
      if (notificationMessage != null)
        'notification_message': notificationMessage,
    };
  }

  /// True se l'evento ha generato una notifica di warning o critica.
  bool get hasAlert => severity != EventSeverity.info;

  static EventSeverity _severityFromString(String value) {
    switch (value) {
      case 'warning':
        return EventSeverity.warning;
      case 'critical':
        return EventSeverity.critical;
      default:
        return EventSeverity.info;
    }
  }

  @override
  String toString() =>
      'GateEvent(id: $id, dir: ${direction.name}, user: $userDisplayName)';

  @override
  bool operator ==(Object other) =>
      identical(this, other) || (other is GateEvent && other.id == id);

  @override
  int get hashCode => id.hashCode;
}

// ── Dati fake per lo sviluppo UI ──────────────────────────────────────
// TODO: rimuovere quando ApiService.getEvents() è connesso al backend.
final List<Map<String, dynamic>> kFakeEventsJson = [
  {
    'id': 'ev-001',
    'timestamp': DateTime.now()
        .subtract(const Duration(minutes: 15))
        .toIso8601String(),
    'direction': 'exit',
    'user_id': 'u-001',
    'user_display_name': 'Simone',
    'rfid_tags': ['E200100860B20004'],
    'object_names': ['MacBook'],
    'severity': 'info',
    'notification_message': null,
  },
  {
    'id': 'ev-002',
    'timestamp': DateTime.now()
        .subtract(const Duration(hours: 2))
        .toIso8601String(),
    'direction': 'exit',
    'user_id': 'u-003',
    'user_display_name': 'Fratello',
    'rfid_tags': [],
    'object_names': [],
    'severity': 'warning',
    'notification_message': 'Fratello è uscito senza telefono',
  },
  {
    'id': 'ev-003',
    'timestamp': DateTime.now()
        .subtract(const Duration(hours: 5))
        .toIso8601String(),
    'direction': 'entry',
    'user_id': 'u-002',
    'user_display_name': 'Mamma',
    'rfid_tags': ['E200100860B20003'],
    'object_names': ['Zaino scuola'],
    'severity': 'info',
    'notification_message': null,
  },
  {
    'id': 'ev-004',
    'timestamp': DateTime.now()
        .subtract(const Duration(hours: 8))
        .toIso8601String(),
    'direction': 'exit',
    'user_id': null,
    'user_display_name': null,
    'rfid_tags': ['E200100860B20001'],
    'object_names': ['Chiavi di casa'],
    'severity': 'critical',
    'notification_message': 'Chiavi uscite senza utente identificato ⚠️',
  },
];
